### ISSUE TITLE
ChainAdapters: Cosmos - implement build delegate tx

### ISSUE THREAD

**Issue by asamere** [shapeshift/lib#330]
The existing buildSendTx function should be enhanced with ability to detect the tx "type" and other required information for building a delegate transaction.

Extend the BuildSendTxInput payload with chainSpecific attributes to accomplish this.  Web will populate the needed information as the user makes selections.

**Comment by NeOMakinG** [comment]
- Why wouldn't we rename `buildSendTransaction` into `buildTransaction`? As it can be Send, Delegate, Redelegate, Undelegate... 

- What do you guys want to add to the `chainSpecific` attribute? We need to pass a sort of `type` attribute in order to detect it, should it be inside chainSpecific?

I played a bit with it, but the implementation's quality is probably really low if you want to take a look for references: https://github.com/shapeshift/lib/commit/a4478631152915e9e81a580a8e4ba1fe03c14352

**Comment by gomesalexandre** [comment]
> * Why wouldn't we rename `buildSendTransaction` into `buildTransaction`? As it can be Send, Delegate, Redelegate, Undelegate...

All chain-adapters are extending the [base `ChainAdapter` type](https://github.com/shapeshift/lib/blob/04b60c544dc275b6ef586248560e91da392a33e7/packages/chain-adapters/src/api.ts#L22) and `buildSendTransaction` comes from that interface, so using another method would kill the abstraction that chain-adapters create.

From a higher-level point-of-view, all a bonding/unbonding is is really just a send Tx.
What a send Tx does (a smart contract call in ETH, some Yearn staking, some Cosmos SDK delegation, etc...) is a more granular concept that isn't really part of describing a send Tx. Cosmos SDK itself defines a Tx as [Tx](https://docs.cosmos.network/master/core/transactions.html#type-definition)/[StdTx](https://docs.cosmos.network/v0.39/modules/auth/03_types.html) in the auth module.

You could argue that delegating and undelegating are protocol-level in Cosmos SDK (well... ish, they're part of the [`staking`](https://docs.cosmos.network/v0.44/modules/staking/01_state.html#delegation) module that a Cosmos SDK blockchain can use), but that really doesn't change the fact that it is still just a Send Tx with `DelegatorAddr` and `ValidatorAddr` info in their message as well as a type.

tl;dr we should be able to use `buildSendTransaction` to build any Tx, including delegations and undelegations. I'm confident that we can do the correct parsing inside `buildSendTransaction()` (as you did in your PoC).

**Comment by elmutt** [comment]
> * Why wouldn't we rename `buildSendTransaction` into `buildTransaction`? As it can be Send, Delegate, Redelegate, Undelegate...
> * What do you guys want to add to the `chainSpecific` attribute? We need to pass a sort of `type` attribute in order to detect it, should it be inside chainSpecific?
> 
> I played a bit with it, the implementation's is wrong but if you want to take a look for references: [a447863](https://github.com/shapeshift/lib/commit/a4478631152915e9e81a580a8e4ba1fe03c14352) (Didn't test it yet, and it's not building anyway)



> > * Why wouldn't we rename `buildSendTransaction` into `buildTransaction`? As it can be Send, Delegate, Redelegate, Undelegate...
> 
> All chain-adapters are extending the [base `ChainAdapter` type](https://github.com/shapeshift/lib/blob/04b60c544dc275b6ef586248560e91da392a33e7/packages/chain-adapters/src/api.ts#L22) and `buildSendTransaction` comes from that interface, so using another method would kill the abstraction that chain-adapters create.
> 
> From a higher-level point-of-view, all a bonding/unbonding is is really just a send Tx. What a send Tx does (a smart contract call in ETH, some Yearn staking, some Cosmos SDK delegation, etc...) is a more granular concept that isn't really part of describing a send Tx. Cosmos SDK itself defines a Tx as [Tx](https://docs.cosmos.network/master/core/transactions.html#type-definition)/[StdTx](https://docs.cosmos.network/v0.39/modules/auth/03_types.html) in the auth module.
> 
> You could argue that delegating and undelegating are protocol-level in Cosmos SDK (well... ish, they're part of the [`staking`](https://docs.cosmos.network/v0.44/modules/staking/01_state.html#delegation) module that a Cosmos SDK blockchain can use), but that really doesn't change the fact that it is still just a Send Tx with `DelegatorAddr` and `ValidatorAddr` info in their message as well as a type.
> 
> tl;dr we should be able to use `buildSendTransaction` to build any Tx, including delegations and undelegations. I'm confident that we can do the correct parsing inside `buildSendTransaction()` (as you did in your PoC).

I think theres a few ways to look at it. 

I see buildSendTransaction as building a normal "send" of one asset from one address to another.  This is how eth and btc are using buildSendTransaction(). Refactoring "send" in the way described would require non-trivial refactoring throughout multiple places in the stack with risk for things to break and I dont see it as that beneficial.

There will always be some features that are specific to certain chain adapters (the eth adapter has an ens related function) -- adding buildWhateverTransaction() adds some code duplication but allows us to make additions to the chain adapters without introducing any real tech debt or dangerous pitfalls.

Check out the linked PR to see the approach we took. I feel like its a pretty future-safe way of doing it but does have some code duplication we can dry up in the future
